File: frontend/src/services/api/forecasting.ts
/**
 * Forecasting API Service
 */

import { apiClient } from './client';

export interface FinancialForecast {
  id: number;
  organization_id: number;
  forecast_name: string;
  forecast_type: string;
  forecast_method: string;
  base_period_start: string;
  base_period_end: string;
  forecast_start: string;
  forecast_end: string;
  frequency: string;
  model_parameters: any;
  business_drivers: any;
  historical_data: any;
  forecast_data: any;
  confidence_intervals?: any;
  accuracy_metrics?: any;
  last_validation_date?: string;
  validation_period_accuracy?: number;
  status: string;
  is_baseline: boolean;
  created_at: string;
  updated_at: string;
  created_by_id?: number;
  approved_by_id?: number;
  approved_at?: string;
}

export interface MLForecastModel {
  id: number;
  organization_id: number;
  model_name: string;
  model_type: string;
  target_variable: string;
  features: string[];
  hyperparameters: any;
  preprocessing_steps: any;
  training_period_start: string;
  training_period_end: string;
  training_data_size: number;
  training_accuracy?: number;
  validation_accuracy?: number;
  test_accuracy?: number;
  mae?: number;
  rmse?: number;
  mape?: number;
  feature_importance?: any;
  model_file_path?: string;
  model_version: string;
  is_active: boolean;
  is_production: boolean;
  trained_at: string;
  created_by_id?: number;
}

export interface MLPrediction {
  id: number;
  ml_model_id: number;
  organization_id: number;
  prediction_date: string;
  predicted_value: number;
  confidence_score?: number;
  lower_bound?: number;
  upper_bound?: number;
  input_features: any;
  actual_value?: number;
  prediction_error?: number;
  predicted_at: string;
}

export interface RiskAnalysis {
  id: number;
  organization_id: number;
  risk_category: string;
  risk_name: string;
  risk_description?: string;
  probability: number;
  impact_score: number;
  risk_score: number;
  potential_financial_impact?: number;
  key_indicators: any;
  threshold_values: any;
  current_indicator_values?: any;
  alert_level: string;
  is_active_alert: boolean;
  mitigation_strategies?: string;
  mitigation_status?: string;
  created_at: string;
  updated_at: string;
  last_assessed_at?: string;
  created_by_id?: number;
}

export interface AutomatedInsight {
  id: number;
  organization_id: number;
  insight_type: string;
  insight_category: string;
  title: string;
  description: string;
  data_sources: string[];
  analysis_method: string;
  confidence_score: number;
  importance_score: number;
  potential_impact?: string;
  recommended_actions?: string;
  supporting_metrics?: any;
  visualization_data?: any;
  status: string;
  user_feedback?: string;
  usefulness_rating?: number;
  generated_at: string;
  expires_at?: string;
  reviewed_by_id?: number;
  reviewed_at?: string;
}

export interface ForecastDashboard {
  active_forecasts: number;
  forecast_accuracy_avg: number;
  recent_predictions: any[];
  risk_alerts: any[];
  key_insights: any[];
  performance_metrics: any;
  trend_analysis: any;
}

// Request types
export interface CreateForecastRequest {
  forecast_name: string;
  forecast_type: string;
  forecast_method: string;
  base_period_start: string;
  base_period_end: string;
  forecast_start: string;
  forecast_end: string;
  frequency: string;
  model_parameters: any;
  business_drivers: any;
  historical_data: any;
}

export interface UpdateForecastRequest {
  forecast_name?: string;
  model_parameters?: any;
  business_drivers?: any;
  status?: string;
  is_baseline?: boolean;
}

export interface CreateMLModelRequest {
  model_name: string;
  model_type: string;
  target_variable: string;
  features: string[];
  hyperparameters: any;
  preprocessing_steps: any;
  training_period_start: string;
  training_period_end: string;
  training_data: any;
}

export interface PredictionRequest {
  model_id: number;
  prediction_date: string;
  input_features: any;
  include_confidence?: boolean;
}

export interface CreateRiskAnalysisRequest {
  risk_category: string;
  risk_name: string;
  risk_description?: string;
  probability: number;
  impact_score: number;
  potential_financial_impact?: number;
  key_indicators: any;
  threshold_values: any;
  mitigation_strategies?: string;
}

export interface UpdateInsightRequest {
  status?: string;
  user_feedback?: string;
  usefulness_rating?: number;
}

// Query parameters
export interface GetForecastsParams {
  skip?: number;
  limit?: number;
  forecast_type?: string;
  forecast_method?: string;
  status?: string;
  is_baseline?: boolean;
}

export interface GetMLModelsParams {
  skip?: number;
  limit?: number;
  model_type?: string;
  target_variable?: string;
  is_active?: boolean;
  is_production?: boolean;
}

export interface GetPredictionsParams {
  model_id?: number;
  start_date?: string;
  end_date?: string;
  skip?: number;
  limit?: number;
}

export interface GetRiskAnalysisParams {
  risk_category?: string;
  alert_level?: string;
  is_active_alert?: boolean;
  skip?: number;
  limit?: number;
}

export interface GetInsightsParams {
  insight_type?: string;
  insight_category?: string;
  status?: string;
  min_importance?: number;
  skip?: number;
  limit?: number;
}

class ForecastingApiService {
  private baseUrl = '/api/v1/forecasting';

  // Financial Forecasts
  async getForecasts(params?: GetForecastsParams): Promise<FinancialForecast[]> {
    const searchParams = new URLSearchParams();
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined) {
          searchParams.append(key, value.toString());
        }
      });
    }
    
    const response = await apiClient.get(
      `${this.baseUrl}/forecasts${searchParams.toString() ? `?${searchParams.toString()}` : ''}`
    );
    return response.data;
  }

  async getForecast(forecastId: number): Promise<FinancialForecast> {
    const response = await apiClient.get(`${this.baseUrl}/forecasts/${forecastId}`);
    return response.data;
  }

  async createForecast(data: CreateForecastRequest): Promise<FinancialForecast> {
    const response = await apiClient.post(`${this.baseUrl}/forecasts`, data);
    return response.data;
  }

  async updateForecast(forecastId: number, data: UpdateForecastRequest): Promise<FinancialForecast> {
    const response = await apiClient.put(`${this.baseUrl}/forecasts/${forecastId}`, data);
    return response.data;
  }

  async deleteForecast(forecastId: number): Promise<void> {
    await apiClient.delete(`${this.baseUrl}/forecasts/${forecastId}`);
  }

  // Business Drivers
  async getBusinessDrivers(forecastId: number, params?: { driver_category?: string; is_active?: boolean }): Promise<any[]> {
    const searchParams = new URLSearchParams();
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined) {
          searchParams.append(key, value.toString());
        }
      });
    }
    
    const response = await apiClient.get(
      `${this.baseUrl}/forecasts/${forecastId}/drivers${searchParams.toString() ? `?${searchParams.toString()}` : ''}`
    );
    return response.data;
  }

  async createBusinessDriver(forecastId: number, data: any): Promise<any> {
    const response = await apiClient.post(`${this.baseUrl}/forecasts/${forecastId}/drivers`, data);
    return response.data;
  }

  // ML Models
  async getMLModels(params?: GetMLModelsParams): Promise<MLForecastModel[]> {
    const searchParams = new URLSearchParams();
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined) {
          searchParams.append(key, value.toString());
        }
      });
    }
    
    const response = await apiClient.get(
      `${this.baseUrl}/ml-models${searchParams.toString() ? `?${searchParams.toString()}` : ''}`
    );
    return response.data;
  }

  async createMLModel(data: CreateMLModelRequest): Promise<MLForecastModel> {
    const response = await apiClient.post(`${this.baseUrl}/ml-models`, data);
    return response.data;
  }

  // Predictions
  async getPredictions(params?: GetPredictionsParams): Promise<MLPrediction[]> {
    const searchParams = new URLSearchParams();
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined) {
          searchParams.append(key, value.toString());
        }
      });
    }
    
    const response = await apiClient.get(
      `${this.baseUrl}/predictions${searchParams.toString() ? `?${searchParams.toString()}` : ''}`
    );
    return response.data;
  }

  async createPrediction(data: PredictionRequest): Promise<MLPrediction> {
    const response = await apiClient.post(`${this.baseUrl}/predictions`, data);
    return response.data;
  }

  // Risk Analysis
  async getRiskAnalyses(params?: GetRiskAnalysisParams): Promise<RiskAnalysis[]> {
    const searchParams = new URLSearchParams();
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined) {
          searchParams.append(key, value.toString());
        }
      });
    }
    
    const response = await apiClient.get(
      `${this.baseUrl}/risk-analysis${searchParams.toString() ? `?${searchParams.toString()}` : ''}`
    );
    return response.data;
  }

  async createRiskAnalysis(data: CreateRiskAnalysisRequest): Promise<RiskAnalysis> {
    const response = await apiClient.post(`${this.baseUrl}/risk-analysis`, data);
    return response.data;
  }

  async updateRiskAnalysis(riskId: number, data: Partial<CreateRiskAnalysisRequest>): Promise<RiskAnalysis> {
    const response = await apiClient.put(`${this.baseUrl}/risk-analysis/${riskId}`, data);
    return response.data;
  }

  // Risk Events
  async createRiskEvent(data: any): Promise<any> {
    const response = await apiClient.post(`${this.baseUrl}/risk-events`, data);
    return response.data;
  }

  // Automated Insights
  async getAutomatedInsights(params?: GetInsightsParams): Promise<AutomatedInsight[]> {
    const searchParams = new URLSearchParams();
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined) {
          searchParams.append(key, value.toString());
        }
      });
    }
    
    const response = await apiClient.get(
      `${this.baseUrl}/insights${searchParams.toString() ? `?${searchParams.toString()}` : ''}`
    );
    return response.data;
  }

  async generateInsights(dataSources: string[], analysisPeriodDays: number = 30): Promise<any> {
    const response = await apiClient.post(`${this.baseUrl}/insights/generate`, null, {
      params: {
        data_sources: dataSources,
        analysis_period_days: analysisPeriodDays
      }
    });
    return response.data;
  }

  async updateInsight(insightId: number, data: UpdateInsightRequest): Promise<AutomatedInsight> {
    const response = await apiClient.put(`${this.baseUrl}/insights/${insightId}`, data);
    return response.data;
  }

  // Dashboard
  async getForecastDashboard(): Promise<ForecastDashboard> {
    const response = await apiClient.get(`${this.baseUrl}/dashboard`);
    return response.data;
  }

  // Advanced Analytics
  async createMultivariateForecast(data: any): Promise<any> {
    const response = await apiClient.post(`${this.baseUrl}/multivariate-forecast`, data);
    return response.data;
  }

  async performSensitivityAnalysis(data: any): Promise<any> {
    const response = await apiClient.post(`${this.baseUrl}/sensitivity-analysis`, data);
    return response.data;
  }

  async getEarlyWarningSignals(params?: { severity?: string; min_confidence?: number }): Promise<any[]> {
    const searchParams = new URLSearchParams();
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (value !== undefined) {
          searchParams.append(key, value.toString());
        }
      });
    }
    
    const response = await apiClient.get(
      `${this.baseUrl}/early-warning-signals${searchParams.toString() ? `?${searchParams.toString()}` : ''}`
    );
    return response.data;
  }

  // Forecast Versions
  async getForecastVersions(forecastId: number): Promise<any[]> {
    const response = await apiClient.get(`${this.baseUrl}/forecasts/${forecastId}/versions`);
    return response.data;
  }

  async analyzeForecastAccuracy(forecastId: number, actualData: any): Promise<any> {
    const response = await apiClient.post(
      `${this.baseUrl}/forecasts/${forecastId}/accuracy-analysis`,
      actualData
    );
    return response.data;
  }
}

export const forecastingApi = new ForecastingApiService();

File: app/schemas/forecasting.py
# app/schemas/forecasting.py
"""
Forecasting Schemas - Request/Response models for ML forecasting and scenario analysis
"""

from pydantic import BaseModel, Field, validator, SkipValidation
from typing import Optional, Dict, List, Any, Union
from datetime import date, datetime
from decimal import Decimal
from enum import Enum

from app.models.forecasting_models import ForecastMethodType, ForecastFrequency, ForecastStatus

# Base forecasting schemas
class FinancialForecastBase(BaseModel):
    forecast_name: str = Field(..., description="Name of the forecast")
    forecast_type: str = Field(..., description="Type of forecast (revenue, expenses, cash_flow, etc.)")
    forecast_method: ForecastMethodType = Field(..., description="Forecasting method")
    base_period_start: SkipValidation[date] = Field(..., description="Historical data period start")
    base_period_end: SkipValidation[date] = Field(..., description="Historical data period end")
    forecast_start: SkipValidation[date] = Field(..., description="Forecast period start")
    forecast_end: SkipValidation[date] = Field(..., description="Forecast period end")
    frequency: ForecastFrequency = Field(..., description="Forecast frequency")
    model_parameters: Dict[str, Any] = Field(..., description="Model parameters and configuration")
    business_drivers: Dict[str, Any] = Field(..., description="Key business drivers")

class FinancialForecastCreate(FinancialForecastBase):
    historical_data: Dict[str, Any] = Field(..., description="Historical data for training")

class FinancialForecastUpdate(BaseModel):
    forecast_name: Optional[str] = None
    model_parameters: Optional[Dict[str, Any]] = None
    business_drivers: Optional[Dict[str, Any]] = None
    status: Optional[ForecastStatus] = None
    is_baseline: Optional[bool] = None

class FinancialForecastResponse(FinancialForecastBase):
    id: int
    organization_id: int
    historical_data: Dict[str, Any]
    forecast_data: Dict[str, Any]
    confidence_intervals: Optional[Dict[str, Any]]
    accuracy_metrics: Optional[Dict[str, Any]]
    last_validation_date: Optional[SkipValidation[date]]
    validation_period_accuracy: Optional[Decimal]
    status: ForecastStatus
    is_baseline: bool
    created_at: datetime
    updated_at: datetime
    created_by_id: Optional[int]
    approved_by_id: Optional[int]
    approved_at: Optional[datetime]

    class Config:
        from_attributes = True

# Business driver schemas
class BusinessDriverBase(BaseModel):
    driver_name: str = Field(..., description="Name of the business driver")
    driver_category: str = Field(..., description="Category (operational, financial, market)")
    driver_type: str = Field(..., description="Type (volume, price, rate, index)")
    impact_formula: str = Field(..., description="Formula for how this driver impacts the forecast")
    elasticity: Optional[Decimal] = Field(None, description="Elasticity coefficient")
    data_source: Optional[str] = Field(None, description="External data source")
    external_api_endpoint: Optional[str] = Field(None, description="API endpoint for external data")

class BusinessDriverCreate(BusinessDriverBase):
    financial_forecast_id: int = Field(..., description="Associated forecast ID")
    historical_values: Dict[str, Any] = Field(..., description="Historical driver values")
    projected_values: Dict[str, Any] = Field(..., description="Projected driver values")

class BusinessDriverUpdate(BaseModel):
    driver_name: Optional[str] = None
    historical_values: Optional[Dict[str, Any]] = None
    projected_values: Optional[Dict[str, Any]] = None
    impact_formula: Optional[str] = None
    elasticity: Optional[Decimal] = None
    is_active: Optional[bool] = None

class BusinessDriverResponse(BusinessDriverBase):
    id: int
    financial_forecast_id: int
    organization_id: int
    historical_values: Dict[str, Any]
    projected_values: Dict[str, Any]
    correlation_metrics: Optional[Dict[str, Any]]
    last_sync_date: Optional[datetime]
    is_active: bool
    created_at: datetime
    updated_at: datetime
    created_by_id: Optional[int]

    class Config:
        from_attributes = True

# ML Model schemas
class MLModelConfiguration(BaseModel):
    model_type: str = Field(..., description="Type of ML model (linear_regression, random_forest, etc.)")
    features: List[str] = Field(..., description="List of input features")
    hyperparameters: Dict[str, Any] = Field(default_factory=dict, description="Model hyperparameters")
    preprocessing_steps: Dict[str, Any] = Field(default_factory=dict, description="Data preprocessing configuration")
    cross_validation_folds: int = Field(default=5, ge=2, le=10, description="Number of CV folds")
    test_size: float = Field(default=0.2, ge=0.1, le=0.5, description="Test set size")

class MLForecastModelBase(BaseModel):
    model_name: str = Field(..., description="ML model name")
    model_type: str = Field(..., description="Type of ML model")
    target_variable: str = Field(..., description="Target variable being predicted")
    features: List[str] = Field(..., description="Input features")
    hyperparameters: Dict[str, Any] = Field(..., description="Model hyperparameters")
    preprocessing_steps: Dict[str, Any] = Field(..., description="Data preprocessing steps")
    training_period_start: SkipValidation[date] = Field(..., description="Training data start date")
    training_period_end: SkipValidation[date] = Field(..., description="Training data end date")

class MLForecastModelCreate(MLForecastModelBase):
    training_data: Dict[str, Any] = Field(..., description="Training dataset")

class MLForecastModelResponse(MLForecastModelBase):
    id: int
    organization_id: int
    training_data_size: int
    training_accuracy: Optional[Decimal]
    validation_accuracy: Optional[Decimal]
    test_accuracy: Optional[Decimal]
    mae: Optional[Decimal]
    rmse: Optional[Decimal]
    mape: Optional[Decimal]
    feature_importance: Optional[Dict[str, Any]]
    model_file_path: Optional[str]
    model_version: str
    is_active: bool
    is_production: bool
    trained_at: datetime
    created_by_id: Optional[int]

    class Config:
        from_attributes = True

# Prediction schemas
class PredictionRequest(BaseModel):
    model_id: int = Field(..., description="ML model ID to use for prediction")
    prediction_date: SkipValidation[date] = Field(..., description="Date to predict for")
    input_features: Dict[str, Any] = Field(..., description="Input feature values")
    include_confidence: bool = Field(default=True, description="Include confidence intervals")

class MLPredictionResponse(BaseModel):
    id: int
    ml_model_id: int
    organization_id: int
    prediction_date: SkipValidation[date]
    predicted_value: Decimal
    confidence_score: Optional[Decimal]
    lower_bound: Optional[Decimal]
    upper_bound: Optional[Decimal]
    input_features: Dict[str, Any]
    actual_value: Optional[Decimal]
    prediction_error: Optional[Decimal]
    predicted_at: datetime

    class Config:
        from_attributes = True

# Risk Analysis schemas
class RiskAnalysisBase(BaseModel):
    risk_category: str = Field(..., description="Risk category (financial, operational, market)")
    risk_name: str = Field(..., description="Name of the risk")
    risk_description: Optional[str] = Field(None, description="Detailed risk description")
    probability: Decimal = Field(..., ge=0, le=100, description="Risk probability (0-100%)")
    impact_score: Decimal = Field(..., ge=1, le=10, description="Risk impact score (1-10)")
    potential_financial_impact: Optional[Decimal] = Field(None, description="Potential financial impact")
    key_indicators: Dict[str, Any] = Field(..., description="Key risk indicators to monitor")
    threshold_values: Dict[str, Any] = Field(..., description="Alert threshold values")
    mitigation_strategies: Optional[str] = Field(None, description="Risk mitigation strategies")

class RiskAnalysisCreate(RiskAnalysisBase):
    pass

class RiskAnalysisUpdate(BaseModel):
    risk_description: Optional[str] = None
    probability: Optional[Decimal] = None
    impact_score: Optional[Decimal] = None
    potential_financial_impact: Optional[Decimal] = None
    key_indicators: Optional[Dict[str, Any]] = None
    threshold_values: Optional[Dict[str, Any]] = None
    current_indicator_values: Optional[Dict[str, Any]] = None
    alert_level: Optional[str] = None
    mitigation_strategies: Optional[str] = None
    mitigation_status: Optional[str] = None

class RiskAnalysisResponse(RiskAnalysisBase):
    id: int
    organization_id: int
    risk_score: Decimal
    current_indicator_values: Optional[Dict[str, Any]]
    alert_level: str
    is_active_alert: bool
    mitigation_status: Optional[str]
    created_at: datetime
    updated_at: datetime
    last_assessed_at: Optional[datetime]
    created_by_id: Optional[int]

    class Config:
        from_attributes = True

# Risk Event schemas
class RiskEventBase(BaseModel):
    event_description: str = Field(..., description="Description of the risk event")
    severity_level: str = Field(..., description="Severity level (low, medium, high, critical)")
    actual_financial_impact: Optional[Decimal] = Field(None, description="Actual financial impact")
    recovery_cost: Optional[Decimal] = Field(None, description="Cost of recovery")
    response_actions: Optional[str] = Field(None, description="Response actions taken")
    lessons_learned: Optional[str] = Field(None, description="Lessons learned from the event")

class RiskEventCreate(RiskEventBase):
    risk_analysis_id: int = Field(..., description="Associated risk analysis ID")
    event_date: SkipValidation[date] = Field(..., description="Date of the risk event")

class RiskEventUpdate(BaseModel):
    event_description: Optional[str] = None
    severity_level: Optional[str] = None
    actual_financial_impact: Optional[Decimal] = None
    recovery_cost: Optional[Decimal] = None
    response_actions: Optional[str] = None
    resolution_date: Optional[SkipValidation[date]] = None
    lessons_learned: Optional[str] = None
    status: Optional[str] = None

class RiskEventResponse(RiskEventBase):
    id: int
    risk_analysis_id: int
    organization_id: int
    event_date: SkipValidation[date]
    resolution_date: Optional[SkipValidation[date]]
    status: str
    created_at: datetime
    updated_at: datetime
    reported_by_id: Optional[int]

    class Config:
        from_attributes = True

# Automated Insights schemas
class AutomatedInsightBase(BaseModel):
    insight_type: str = Field(..., description="Type of insight (trend, anomaly, opportunity, risk)")
    insight_category: str = Field(..., description="Category (financial, operational, market)")
    title: str = Field(..., description="Insight title")
    description: str = Field(..., description="Detailed insight description")
    analysis_method: str = Field(..., description="Analysis method used")
    confidence_score: Decimal = Field(..., ge=0, le=100, description="Confidence score (0-100%)")
    importance_score: Decimal = Field(..., ge=1, le=10, description="Importance score (1-10)")
    potential_impact: Optional[str] = Field(None, description="Potential impact description")
    recommended_actions: Optional[str] = Field(None, description="Recommended actions")

class AutomatedInsightCreate(AutomatedInsightBase):
    data_sources: List[str] = Field(..., description="Data sources analyzed")
    supporting_metrics: Optional[Dict[str, Any]] = Field(None, description="Supporting metrics")
    visualization_data: Optional[Dict[str, Any]] = Field(None, description="Data for visualizations")
    expires_at: Optional[datetime] = Field(None, description="When insight becomes stale")

class AutomatedInsightUpdate(BaseModel):
    status: Optional[str] = None
    user_feedback: Optional[str] = None
    usefulness_rating: Optional[int] = Field(None, ge=1, le=5, description="Usefulness rating (1-5)")

class AutomatedInsightResponse(AutomatedInsightBase):
    id: int
    organization_id: int
    data_sources: List[str]
    supporting_metrics: Optional[Dict[str, Any]]
    visualization_data: Optional[Dict[str, Any]]
    status: str
    user_feedback: Optional[str]
    usefulness_rating: Optional[int]
    generated_at: datetime
    expires_at: Optional[datetime]
    reviewed_by_id: Optional[int]
    reviewed_at: Optional[datetime]

    class Config:
        from_attributes = True

# Forecast version schemas
class ForecastVersionResponse(BaseModel):
    id: int
    financial_forecast_id: int
    organization_id: int
    version_number: str
    version_name: Optional[str]
    change_description: Optional[str]
    forecast_snapshot: Dict[str, Any]
    model_parameters_snapshot: Dict[str, Any]
    is_current: bool
    created_at: datetime
    created_by_id: Optional[int]

    class Config:
        from_attributes = True

# Complex forecasting requests
class MultiVariateForecastRequest(BaseModel):
    forecast_name: str = Field(..., description="Name for the forecast")
    target_variables: List[str] = Field(..., description="Variables to forecast")
    predictor_variables: List[str] = Field(..., description="Predictor variables")
    forecast_horizon: int = Field(..., ge=1, le=60, description="Forecast horizon in periods")
    frequency: ForecastFrequency = Field(..., description="Forecast frequency")
    model_type: str = Field(default="auto", description="Model type (auto, var, lstm, etc.)")
    include_seasonality: bool = Field(default=True, description="Include seasonal components")
    include_trend: bool = Field(default=True, description="Include trend components")
    confidence_level: float = Field(default=0.95, ge=0.8, le=0.99, description="Confidence level for intervals")

class ScenarioForecastRequest(BaseModel):
    base_forecast_id: int = Field(..., description="Base forecast to create scenarios from")
    scenarios: List[Dict[str, Any]] = Field(..., description="List of scenario definitions")
    include_monte_carlo: bool = Field(default=False, description="Include Monte Carlo simulation")
    monte_carlo_iterations: int = Field(default=1000, ge=100, le=10000, description="Monte Carlo iterations")

class SensitivityAnalysisRequest(BaseModel):
    forecast_id: int = Field(..., description="Forecast to analyze")
    sensitivity_variables: List[str] = Field(..., description="Variables to test sensitivity for")
    variation_range: Dict[str, float] = Field(..., description="Variation range for each variable (±%)")
    steps: int = Field(default=10, ge=5, le=50, description="Number of steps in sensitivity analysis")

class ForecastAccuracyReport(BaseModel):
    forecast_id: int
    actual_vs_predicted: Dict[str, Any]
    accuracy_metrics: Dict[str, Decimal]
    error_analysis: Dict[str, Any]
    recommendations: List[str]
    analysis_period: str
    generated_at: datetime

class EarlyWarningSignal(BaseModel):
    signal_type: str = Field(..., description="Type of warning signal")
    severity: str = Field(..., description="Severity level (low, medium, high, critical)")
    description: str = Field(..., description="Signal description")
    affected_metrics: List[str] = Field(..., description="Affected business metrics")
    threshold_breach: Dict[str, Any] = Field(..., description="Threshold breach details")
    recommended_actions: List[str] = Field(..., description="Recommended immediate actions")
    confidence_score: Decimal = Field(..., description="Confidence in the signal")
    triggered_at: datetime = Field(..., description="When the signal was triggered")

class ForecastDashboardData(BaseModel):
    active_forecasts: int
    forecast_accuracy_avg: Decimal
    recent_predictions: List[Dict[str, Any]]
    risk_alerts: List[Dict[str, Any]]
    key_insights: List[Dict[str, Any]]
    performance_metrics: Dict[str, Any]
    trend_analysis: Dict[str, Any]